ssl/ech/ech_internal.c

```diff
@@ -160,4 +227,1223 @@ int ossl_ech_conn_init(SSL_CONNECTION *s, SSL_CTX *ctx,
+    size_t retlen = 0, labellen = 0;
+    EVP_PKEY_CTX *pctx = NULL;
+    const char *label = NULL;
+    unsigned char *p = NULL;
+
+    if (for_hrr == 1) {
+        label = OSSL_ECH_HRR_CONFIRM_STRING;
+    } else {
+        label = OSSL_ECH_ACCEPT_CONFIRM_STRING;
+    }
+    labellen = strlen(label);
+# ifdef OSSL_ECH_SUPERVERBOSE
+    ech_pbuf("cc: label", (unsigned char *)label, labellen);
+# endif
+    memset(zeros, 0, EVP_MAX_MD_SIZE);
+    /* We don't seem to have an hkdf-extract that's exposed by libcrypto */
+    pctx = EVP_PKEY_CTX_new_id(EVP_PKEY_HKDF, NULL);
+    if (pctx == NULL
+        || EVP_PKEY_derive_init(pctx) != 1
+        || EVP_PKEY_CTX_hkdf_mode(pctx,
+                                  EVP_PKEY_HKDEF_MODE_EXTRACT_ONLY) != 1
+        || EVP_PKEY_CTX_hkdf_mode(pctx,
+                                  EVP_PKEY_HKDEF_MODE_EXTRACT_ONLY) != 1
+        || EVP_PKEY_CTX_set_hkdf_md(pctx, md) != 1) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    /* pick correct client_random */
+    if (s->server)
+        p = s->s3.client_random;
+    else
+        p = s->ext.ech.client_random;
+# ifdef OSSL_ECH_SUPERVERBOSE
+    ech_pbuf("cc: client_random", p, SSL3_RANDOM_SIZE);
+# endif
+    if (EVP_PKEY_CTX_set1_hkdf_key(pctx, p, SSL3_RANDOM_SIZE) != 1
+        || EVP_PKEY_CTX_set1_hkdf_salt(pctx, zeros, hashlen) != 1
+        || EVP_PKEY_derive(pctx, NULL, &retlen) != 1
+        || hashlen != retlen
+        || EVP_PKEY_derive(pctx, notsecret, &retlen) != 1) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+# ifdef OSSL_ECH_SUPERVERBOSE
+    ech_pbuf("cc: notsecret", notsecret, hashlen);
+# endif
+    if (hashlen < 8
+        || !tls13_hkdf_expand(s, md, notsecret,
+                              (const unsigned char *)label, labellen,
+                              hashval, hashlen, hoval, 8, 1)) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    rv = 1;
+err:
+    EVP_PKEY_CTX_free(pctx);
+    return rv;
+}
+
+/*
+ * ECH accept_confirmation calculation
+ * for_hrr is 1 if this is for an HRR, otherwise for SH
+ * ac is (a caller allocated) 8 octet buffer
+ * shbuf is a pointer to the SH buffer (incl. the type+3-octet length)
+ * shlen is the length of the SH buf
+ * return: 1 for success, 0 otherwise
+ *
+ * This is a magic value in the ServerHello.random lower 8 octets
+ * that is used to signal that the inner worked.
+ *
+ * As per spec:
+ *
+ * accept_confirmation = HKDF-Expand-Label(
+ *         HKDF-Extract(0, ClientHelloInner.random),
+ *         "ech accept confirmation",
+ *         transcript_ech_conf,
+ *         8)
+ *
+ * transcript_ech_conf = ClientHelloInner..ServerHello
+ *         with last 8 octets of ServerHello.random==0x00
+ *
+ * and with differences due to HRR
+ */
+int ech_calc_confirm(SSL_CONNECTION *s, int for_hrr, unsigned char *acbuf,
```

Comment:
You may write `unsigned char acbuf[8],` , so compiler will be able to do more policy.

Reply:
really? I'd have expected some CI cross-compile might hate that, but sure, let's give it a try...